logfire/_internal/exporters/tail_sampling.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+import random
+from dataclasses import dataclass
+from functools import cached_property
+from threading import Lock
+
+from opentelemetry import context
+from opentelemetry.sdk.trace import ReadableSpan, Span, SpanProcessor
+
+from logfire._internal.constants import (
+    ATTRIBUTES_LOG_LEVEL_NUM_KEY,
+    LEVEL_NUMBERS,
+    ONE_SECOND_IN_NANOSECONDS,
+    LevelName,
+)
+from logfire._internal.exporters.wrapper import WrapperSpanProcessor
+
+
+@dataclass
+class TailSamplingOptions:
+    level: LevelName | None = 'notice'
+    """
+    Include all spans/logs with level greater than or equal to this level.
+    If None, spans are not included based on level.
+    """
+
+    duration: float | None = 1.0
+    """
+    Include all spans/logs with duration greater than this duration in seconds.
+    If None, spans are not included based on duration.
+    """
+
+
+@dataclass
+class TraceBuffer:
+    """Arguments of `on_start` and `on_end` for spans in a single trace."""
+
+    started: list[tuple[Span, context.Context | None]]
+    ended: list[ReadableSpan]
+
+    @cached_property
+    def first_span(self) -> Span:
+        return self.started[0][0]
+
+
+class TailSamplingProcessor(WrapperSpanProcessor):
+    """Passes spans to the wrapped processor if any span in a trace meets the sampling criteria."""
+
+    def __init__(self, processor: SpanProcessor, options: TailSamplingOptions, random_rate: float) -> None:
+        super().__init__(processor)
+        self.duration: float = (
+            float('inf') if options.duration is None else options.duration * ONE_SECOND_IN_NANOSECONDS
+        )
```

Comment:
I'd like these durations to have the unit in their name. So `self.duration` -> `self.duration_ns` and `TailsamplingOptions.duration` -> `TailSamplingOptions.duration_sec` or something like that.